lib/Dialects/VectorExt/Transform/VectorMaskingUtils.cpp

```diff
@@ -0,0 +1,138 @@
+#include "mlir/IR/AffineMap.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallPtrSet.h"
+
+using namespace mlir;
+using namespace mlir::linalg;
+using namespace mlir::vector;
+using namespace mlir::vector_ext;
+
+using ExprList = ArrayRef<ArrayRef<AffineExpr>>;
+
+/// Move the operation range `opRange` before operation `dest`.
+static void moveOperationsBefore(llvm::iterator_range<Block::iterator> opRange,
+                                 Operation *dest) {
+  if (opRange.empty())
+    return;
+  dest->getBlock()->getOperations().splice(
+      Block::iterator(dest), opRange.begin()->getBlock()->getOperations(),
+      opRange.begin(), opRange.end());
+}
+
+/// Iterate over all the definitions produced by the operation range `opRange`
+/// and gather those with users outside the range.
+static void
+getDefsWithExternalUses(llvm::iterator_range<Block::iterator> opRange,
+                        SmallVectorImpl<Value> &defsWithExternalUses,
+                        SmallVectorImpl<Type> &defTypes) {
+  SmallPtrSet<Operation *, 8> rangeSet;
+  for (Operation &op : opRange)
+    rangeSet.insert(&op);
+
+  for (Operation &op : llvm::make_early_inc_range(opRange))
+    for (Value res : op.getResults())
+      for (Operation *user : res.getUsers())
+        if (rangeSet.count(user) == 0) {
+          defsWithExternalUses.push_back(res);
+          defTypes.push_back(res.getType());
+          break;
+        }
+}
+
+/// Introduce a vector.predicate op that encloses the loop body of the tiled
+/// loop. The predicate is computed based on the provided vectorization factor.
+static void createVectorPredicateOp(TiledLoopOp loopOp, int64_t vecFactor) {
+  // Compute the range of operations that will be moved within vector.predicate
+  // and the definitions within the range with external users.
+  Block &bodyBlock = loopOp.getLoopBody().front();
+  auto opsToMove =
+      llvm::make_range(bodyBlock.begin(), Block::iterator(bodyBlock.back()));
+
+  SmallVector<Value, 8> defsWithExternalUses;
+  SmallVector<Type, 8> defTypes;
+  getDefsWithExternalUses(opsToMove, defsWithExternalUses, defTypes);
+
+  // Initialize a builder to insert vector.predicate at the beginning of the
+  // loop body.
+  Location loc = loopOp.getLoc();
+  MLIRContext *context = loopOp.getContext();
+  OpBuilder builder(loopOp);
+  builder.setInsertionPointToStart(&bodyBlock);
+
+  // Generate the predicate to used by the vector.predicate operation:
+  //   %min = affine.min affine_map<(d0)[s0] -> (vecFactor, -d0 + s0)>(%iv)[%ub]
+  //   %vec_pred = vector.create_mask %min : vector<8xi1>
+  // TODO: This is probably not the most efficient way to generate the vector
+  // predicate. Consider using the vector IV.
+  AffineExpr i, j;
+  bindDims(context, i, j);
+  SmallVector<AffineMap, 4> maps = AffineMap::inferFromExprList(
+      ExprList{{builder.getAffineConstantExpr(vecFactor), i - j}});
+  auto minOp = builder.create<AffineMinOp>(
```

Comment:
use/refactor/extend the AffineBuilder utils?
e.g. https://github.com/google/iree-llvm-sandbox/blob/main/lib/Dialects/LinalgExt/Transforms/TileToSequentialFor.cpp#L73

Reply:
Good point. I didn't know about that builder! However, AffineBuilder is part of the sandbox so I think it's better to avoid dependences with other pieces of the sandbox so that we can facilitate the graduation/reversion of the experimental code (as we described in the sandbox guidelines). Does that make sense? Or  do you plan to move that to core in the upcoming days?

Reply:
There is an ArithBuilder with similar purpose.
We can move it next to it.

Ideally I'd wait for affine_apply / min / max to move to arith but it will likely take time: https://llvm.discourse.group/t/rfc-arith-dialect-versions-of-affine-apply-min-max/4656

But if my choice is between using verbose APIs and moving to core before it is in the right place, I'd bias towards moving to core :)